### ISSUE TITLE
Issues loading a .pb file generated using python

### ISSUE THREAD

**Issue by DinoSourcesRex** [migueldeicaza/TensorFlowSharp#316]
**Describe the bug**
I receive an `Invalid Graphdef` exception when attempting to import a pb file I have successfully saved, and loaded, using python.

I've created a [repo ](https://github.com/DinoSourcesRex/TensorflowSharpLoadingAModel)with some code to get you started. Just run the application. The python scripts are there too.

**To Reproduce**
Please see the [repo](https://github.com/DinoSourcesRex/TensorflowSharpLoadingAModel)

The build steps are in the readme, open the sln and hit `F5` and you should be up and running.

Get back to me here if that somehow doesn't work. (make sure you're targetting x64).

**Expected behavior**
I expect the model to load.

Failing that instructions / examples on how to save a model in python to get it to work with TensorflowSharp would be ideal.

**Desktop (please complete the following information):**
 - OS: Windows 10
 - Version	10.0.17134 Build 17134

**Comment by DinoSourcesRex** [comment]
A minor update to this. I can load the `inception.pb` graph with no issues. How was this packaged? I am using the [simple_save](https://www.tensorflow.org/api_docs/python/tf/saved_model/simple_save) function in my python scripts.

If anyone can point me in the direction of which method to use when saving my python model that would be very helpful.

**Comment by ibrcic** [comment]
Did you ever resolve this? I have very similar problem as you.

**Comment by DinoSourcesRex** [comment]
@ibrcic No, it is not resolved. A co-worker and I simply decided we'd be better off going down the python route even though none of us have experience in Python commercially.

I'm hoping that at some point someone who knows the library writes a few examples on how to load these models as there are a lot of issues in the past about trouble loading models, and it is definitely one of the first things people will want to do.

**Comment by DinoSourcesRex** [comment]
@ibrcic No, it is not resolved. A co-worker and I simply decided we'd be better off going down the python route even though none of us have experience in Python.

I'm hoping that at some point someone who knows the library writes a few examples on how to load these models as there are a lot of issues in the past about trouble loading models, and it is definitely one of the first things people will want to do.

**Comment by Shivanitiwary** [comment]
You get this error because the graph is not frozen properly. Use [freeze_graph](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/tools/freeze_graph.py) for tensorflow graph. You should know the input and output node names for the model to freeze and to infer. You can use something like [this](https://stackoverflow.com/questions/36883949/in-tensorflow-get-the-names-of-all-the-tensors-in-a-graph) to know the names. For me its working perfectly fine using proper frozen graph.

**Comment by DinoSourcesRex** [comment]
@Shivanitiwary As far as I am aware using `saved_model` is the suggested way of saving a model. To say it is not being saved correctly is not quite correct since that's what is currently on the [Tensorflow docs](https://www.tensorflow.org/guide/saved_model) for saving and loading.

**Comment by Shivanitiwary** [comment]
@DinoSourcesRex Its not TensorSharp that is unable to use saved_model but in general to deploy a model into any tool you need all the checkpoints, meta information and GraphDefs frozen together in the .protobuf file as explained in [here](https://www.tensorflow.org/mobile/prepare_models). I hope the given links will help you understand it clearly.

**Comment by DinoSourcesRex** [comment]
@Shivanitiwary Currently I am deploying my model, saved with `simple_save`, via Tensorflow Serving. It works fine the way it is.

My question is whether there are some technical limitations stopping TensorflowSharp from doing the same. It would increase adoption of the library as well as drastically reduce the number of issues raised over loading models, I am sure.

**Comment by Shivanitiwary** [comment]
@DinoSourcesRex  Thanks for letting me know about simple_save. I didn't know about it and I would like to try it too. So, when I had the error of `Invalid Graphdef ` I solved it using by freeze_graph.py. I have a question so you saved the model in python using simple_save and now you can deploy your model in TensorFlowSharp correct? 
And if it is working then I can only say that you can use any method, which is available to correctly save the GraphDef, meta information and checkpoints together and you will be able to use it in TensorflowSharp or any other app. As far as my understanding TensorflowSharp has no limitations, it just need all the information in one protobuf file. And the  `Invalid Graphdef `  error occurs only because there is something missing in protobuf file.  
I hope, it helps.

**Comment by DinoSourcesRex** [comment]
@Shivanitiwary Here is my current process:

1. Export my graph using `simple_save`
2. Create a versioned directory (something like 000001, 000002 etc) for each model version
3. I personally use the docker version of Tensorflow Serving `docker run --rm -it -p 8501:8501 -v /c/modelversions/:/models/model_name -e MODEL_NAME=model_name -t tensorflow/serving` where the local path is the folder which contains all the versioned folders of the `.pb` files

The only downside here is, because I do onehot encoding, I have to send pre-encoded arrays. It's still a WIP in terms of figuring out how it works.

---

Regardless the takeaway is that using `simple_save` gives me a folder with everything I need to serve the models immediately in Tensorflow Serving. It worked immediately.

With TensorflowSharp when I try to load the same `.pb` file the graph throws an `Invalid Graphdef` - I personally do not think this should happen, TensorflowSharp should support this format. I had no issues in Python or Tensorflow Serving.

I will probably try to freeze the model in the manner you've suggested to sate my curiosity however for production, which this code is for, I simply wouldn't be able to commit my team to using TensorflowSharp over something like Tensorflow Serving which is a shame as we're all .NET developers and would much prefer to work with a language we understand.
